File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/photos/PhotoDetailViewModel.java
package com.eulersbridge.isegoria.feed.photos;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import com.eulersbridge.isegoria.IsegoriaApp;
import com.eulersbridge.isegoria.network.api.models.Like;
import com.eulersbridge.isegoria.network.api.models.Photo;
import com.eulersbridge.isegoria.network.api.models.User;
import com.eulersbridge.isegoria.network.api.responses.LikeResponse;
import com.eulersbridge.isegoria.util.data.RetrofitLiveData;
import com.eulersbridge.isegoria.util.data.SingleLiveData;

import java.util.List;

@SuppressWarnings("WeakerAccess")
public class PhotoDetailViewModel extends AndroidViewModel {

    private List<Photo> photos;
    private final MutableLiveData<Integer> visibleIndex = new MutableLiveData<>();

    final LiveData<Photo> currentPhoto = Transformations.switchMap(visibleIndex, index -> {
        if (photos != null && index >= 0 && index < photos.size()) {
            return new SingleLiveData<>(photos.get(index));
        } else {
            return new SingleLiveData<>(null);
        }
    });

    final LiveData<List<Like>> photoLikes = Transformations.switchMap(currentPhoto, photo -> {
        IsegoriaApp isegoriaApp = getApplication();

        final long photoId = photo.id;

        LiveData<List<Like>> photoLikes = new RetrofitLiveData<>(isegoriaApp.getAPI().getPhotoLikes(photoId));

        return Transformations.switchMap(photoLikes, likes -> {
            if (currentPhoto.getValue() != null && photoId == currentPhoto.getValue().id) {
                return new SingleLiveData<>(likes);

            } else {
                return new SingleLiveData<>(null);
            }
        });
    });

    final LiveData<Boolean> photoLikedByUser = Transformations.switchMap(photoLikes, likes -> {
        if (likes != null) {
            IsegoriaApp isegoriaApp = getApplication();
            User user = isegoriaApp.loggedInUser.getValue();

            if (user != null)
                for (Like like : likes)
                    if (like.email.equals(user.email))
                        return new SingleLiveData<>(true);
        }

        return new SingleLiveData<>(false);
    });

    public PhotoDetailViewModel(@NonNull Application application) {
        super(application);
    }

    @Override
    protected void onCleared() {
        if (photoLikes instanceof RetrofitLiveData)
            ((RetrofitLiveData) photoLikes).cancel();
    }

    void changePhoto(@IntRange(from = 0) int position) {

        /* Cancel fetching likes for old photo if started,
            but give like/unlike requests a chance to finish */
        if (photoLikes instanceof RetrofitLiveData)
            ((RetrofitLiveData) photoLikes).cancel();


        visibleIndex.setValue(position);
    }

    void setPhotos(List<Photo> photos, int startIndex) {
        this.photos = photos;
        visibleIndex.setValue(startIndex);
    }

    int getPhotoCount() {
        return (photos == null)? 0 : photos.size();
    }

    /**
     * @return Boolean.TRUE on success, Boolean.FALSE on failure
     */
    LiveData<Boolean> likePhoto() {
        final IsegoriaApp isegoriaApp = getApplication();

        final User user = isegoriaApp.loggedInUser.getValue();

        if (user != null) {
            return Transformations.switchMap(currentPhoto, article -> {
                LiveData<LikeResponse> newsArticleLike = new RetrofitLiveData<>(isegoriaApp.getAPI().likePhoto(article.id, user.email));

                return Transformations.switchMap(newsArticleLike,
                        likeResponse -> new SingleLiveData<>(likeResponse != null && likeResponse.success));
            });
        }

        return new SingleLiveData<>(false);
    }

    /**
     * @return Boolean.TRUE on success, Boolean.FALSE on failure
     */
    LiveData<Boolean> unlikePhoto() {
        IsegoriaApp isegoriaApp = getApplication();

        User user = isegoriaApp.loggedInUser.getValue();

        if (user != null) {
            return Transformations.switchMap(currentPhoto,
                    photo -> {
                        LiveData<Void> unlike = new RetrofitLiveData<>(isegoriaApp.getAPI().unlikePhoto(photo.id, user.email));

                        return Transformations.switchMap(unlike, __ -> new SingleLiveData<>(true));
                    });
        }

        return new SingleLiveData<>(false);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/photos/PhotoDetailViewModel.kt
package com.eulersbridge.isegoria.feed.photos

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.support.annotation.IntRange
import com.eulersbridge.isegoria.IsegoriaApp
import com.eulersbridge.isegoria.network.api.models.Like
import com.eulersbridge.isegoria.network.api.models.Photo
import com.eulersbridge.isegoria.util.data.RetrofitLiveData
import com.eulersbridge.isegoria.util.data.SingleLiveData

class PhotoDetailViewModel(application: Application) : AndroidViewModel(application) {

    private var photos: List<Photo>? = null
    private val visibleIndex = MutableLiveData<Int>()

    internal val currentPhoto = Transformations.switchMap<Int, Photo>(visibleIndex) { index ->
        SingleLiveData(photos?.get(index))
    }

    internal fun getPhotoUrl(position: Int): String? = photos?.get(position)?.thumbnailUrl

    private fun getPhotoLikes(): LiveData<List<Like>> {
        return Transformations.switchMap(currentPhoto) { (id) ->
            val app = getApplication<IsegoriaApp>()
            val photoId = id.toLong()
            val photoLikes = RetrofitLiveData(app.api.getPhotoLikes(photoId))

            Transformations.switchMap<List<Like>, List<Like>>(photoLikes) likes@{ likes ->
                if (photoId == currentPhoto.value?.id?.toLong()) {
                    return@likes SingleLiveData(likes)

                } else {
                    return@likes SingleLiveData(null)
                }
            }
        }
    }

    internal fun getPhotoLikeCount(): LiveData<Int> {
        return Transformations.switchMap(getPhotoLikes()) { likes ->
            SingleLiveData(likes?.size ?: 0)
        }
    }

    internal fun getPhotoLikedByUser(): LiveData<Boolean> {
        return Transformations.switchMap(getPhotoLikes()) { likes ->
            val app = getApplication<IsegoriaApp>()
            val user = app.loggedInUser.value

            if (user != null) {
                val like = likes?.singleOrNull {
                    it.email == user.email
                }

                if (like != null)
                    SingleLiveData(true)
            }

            SingleLiveData(false)
        }
    }

    internal val photoCount: Int by lazy {
        photos?.size ?: 0
    }

    internal fun changePhoto(@IntRange(from = 0) position: Int) {
        visibleIndex.value = position
    }

    internal fun setPhotos(photos: List<Photo>, startIndex: Int) {
        this.photos = photos
        visibleIndex.value = startIndex
    }

    /**
     * @return Boolean.TRUE on success, Boolean.FALSE on failure
     */
    internal fun likePhoto(): LiveData<Boolean> {
        val app = getApplication<IsegoriaApp>()

        app.loggedInUser.value?.let { user ->
            return Transformations.switchMap(currentPhoto) { (id) ->
                val newsArticleLike =
                    RetrofitLiveData(app.api.likePhoto(id.toLong(), user.email))

                Transformations.switchMap(
                    newsArticleLike
                ) { likeResponse -> SingleLiveData(likeResponse != null && likeResponse.success) }
            }
        }

        return SingleLiveData(false)
    }

    /**
     * @return Boolean.TRUE on success, Boolean.FALSE on failure
     */
    internal fun unlikePhoto(): LiveData<Boolean> {
        val app = getApplication<IsegoriaApp>()

        app.loggedInUser.value?.let { user ->
            return Transformations.switchMap(
                currentPhoto
            ) { (id) ->
                val unlike = RetrofitLiveData(app.api.unlikePhoto(id.toLong(), user.email))

                Transformations.switchMap(unlike) { SingleLiveData(true) }
            }
        }

        return SingleLiveData(false)
    }
}
